open_flamingo/train/losses.py

```diff
@@ -0,0 +1,90 @@
+from open_flamingo.src.vlm import VLM
+import torch
+
+SUPPORTED_LOSSES = ["next_token_prediction"]
+
+
+def get_loss_fn(loss_name):
+    if loss_name == "next_token_prediction":
+        return NextTokenPrediction()
+    else:
+        raise ValueError(
+            f"Loss {loss_name} not supported. Supported losses: {SUPPORTED_LOSSES}"
+        )
+
+
+class Loss:
+    @property
+    def name(self):
+        raise NotImplementedError
+
+    def __call__(
+        self,
+        model: VLM,
+        tokenizer,
+        images: torch.Tensor,
+        input_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        autocast: callable,
+    ):
+        """
+        Args:
+            model: VLM model
+            images: images tensor, already moved to device and cast to appropriate dtype
+                shape (B, T_img, F, C, H, W)
+            input_ids: input ids tensor, already moved to device and cast to appropriate dtype
+                shape (B, T_text)
+            attention_mask: attention mask tensor, already moved to device and cast to appropriate dtype
+                shape (B, T_text)
+            autocast: autocast context manager
+        Return:
+            loss: scalar loss
+        """
+        raise NotImplementedError
+
+
+class NextTokenPrediction(Loss):
```

Comment:
I feel we should decouple model forward pass with loss implementation and only pass it logits, labels, etc. here? IMO the benefit of having multiple losses is being able to apply multiple of them at the same time. For instance I am adding z-loss and we want to apply that and next token prediction loss for a single forward pass. This would also match architectures that use multiple losses like CoCa and MaMMUT.

Reply:
ok actually z-loss + cross entropy could just be its own loss.

Reply:
Okay! I was thinking about MaMMUT when structuring it this way; I thought that they required two forward passes of the decoder (one without vision xattn). Open to changing in the future though, if one forward pass ends up being the correct design.